analyzers/tests/SonarAnalyzer.UnitTest/Rules/Utilities/TokenTypeAnalyzerTest.ClassifierTestHarness.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.Text;
+using Moq;
+using SonarAnalyzer.Protobuf;
+using Match = System.Text.RegularExpressions.Match;
+
+namespace SonarAnalyzer.UnitTest.Rules;
+
+public partial class TokenTypeAnalyzerTest
+{
+    private static class ClassifierTestHarness
+    {
+        private const int TokenAnnotationChars = 4; // [u:]
+        private const int PrefixTokenAnnotationChars = 3; // [u:
+
+        public static void AssertTokenTypes(string code, bool allowSemanticModel = true, bool ignoreCompilationErrors = false)
+        {
+            var (tree, model, expectedTokens) = ParseTokens(code, ignoreCompilationErrors);
+            var root = tree.GetRoot();
+            model = allowSemanticModel ? model : new Mock<SemanticModel>(MockBehavior.Strict).Object; // The Mock will throw if the semantic model was used.
+            var tokenClassifier = new SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer.TokenClassifier(model, false);
+            var triviaClassifier = new SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer.TriviaClassifier();
+            expectedTokens.Should().SatisfyRespectively(expectedTokens.Select((Func<ExpectedToken, Action<ExpectedToken>>)(e => CheckClassifiedToken)));
+
+            void CheckClassifiedToken(ExpectedToken expected)
+            {
+                var expectedLineSpan = tree.GetLocation(expected.Position).GetLineSpan();
+                var because = $$"""token with text "{{expected.TokenText}}" at position {{expectedLineSpan}} was marked as {{expected.TokenType}}""";
+                var (actualLocation, classification) = FindActual();
+                if (classification == null)
+                {
+                    because = $$"""classification for token with text "{{expected.TokenText}}" at position {{expectedLineSpan}} is null""";
+                    expected.TokenType.Should().Be(TokenType.UnknownTokentype, because);
+                    actualLocation.SourceSpan.Should().Be(expected.Position, because);
+                }
+                else
+                {
+                    classification.Should().Be(new TokenTypeInfo.Types.TokenInfo
+                    {
+                        TokenType = expected.TokenType,
+                        TextRange = new TextRange
+                        {
+                            StartLine = expectedLineSpan.StartLinePosition.Line + 1,
+                            StartOffset = expectedLineSpan.StartLinePosition.Character,
+                            EndLine = expectedLineSpan.EndLinePosition.Line + 1,
+                            EndOffset = expectedLineSpan.EndLinePosition.Character,
+                        },
+                    }, because);
+                }
+
+                (Location, TokenTypeInfo.Types.TokenInfo TokenInfo) FindActual()
```

Comment:
I would move also this one outside. I find it harder to read nested methods.

Reply:
It captures over:
* expected
* root
* tree
* triviaClassifier
* because

Do we really want to pass in all these values as parameters?

Reply:
I think not all these parameters are needed. TriviaClassifier can be a shared field. It doesn't have any state. Also you can easily get the root if you have the tree. Then the methods can be simplified a bit:
```
public static void AssertTokenTypes(string code, bool allowSemanticModel = true, bool ignoreCompilationErrors = false)
        {
            var (tree, model, expectedTokens) = ParseTokens(code, ignoreCompilationErrors);
            model = allowSemanticModel ? model : new Mock<SemanticModel>(MockBehavior.Strict).Object; // The Mock will throw if the semantic tokenClassifier was used.
            var tokenClassifier = new TokenTypeAnalyzer.TokenClassifier(model, false);
            expectedTokens.Should().SatisfyRespectively(expectedTokens.Select((Func<ExpectedToken, Action<ExpectedToken>>)(_ => token => CheckClassifiedToken(token, tree, tokenClassifier))));
        }
```
```
private static void CheckClassifiedToken(ExpectedToken expected, SyntaxTree tree, TokenTypeAnalyzer.TokenClassifier tokenClassifier)
        {
            var expectedLineSpan = tree.GetLocation(expected.Position).GetLineSpan();
            var because = $$"""token with text "{{expected.TokenText}}" at position {{expectedLineSpan}} was marked as {{expected.TokenType}}""";
            var (actualLocation, classification) = FindActual(expected, tokenClassifier, because, tree);
            if (classification == null)
            {
                because = $$"""classification for token with text "{{expected.TokenText}}" at position {{expectedLineSpan}} is null""";
                expected.TokenType.Should().Be(TokenType.UnknownTokentype, because);
                actualLocation.SourceSpan.Should().Be(expected.Position, because);
            }
            else
            {
                classification.Should().Be(new TokenTypeInfo.Types.TokenInfo
                {
                    TokenType = expected.TokenType,
                    TextRange = new TextRange
                    {
                        StartLine = expectedLineSpan.StartLinePosition.Line + 1,
                        StartOffset = expectedLineSpan.StartLinePosition.Character,
                        EndLine = expectedLineSpan.EndLinePosition.Line + 1,
                        EndOffset = expectedLineSpan.EndLinePosition.Character,
                    },
                }, because);
            }
        }
```
```
private static (Location, TokenTypeInfo.Types.TokenInfo TokenInfo) FindActual(ExpectedToken expected, TokenTypeAnalyzer.TokenClassifier tokenClassifier, string because, SyntaxTree tree)
        {
            if (expected.TokenType == TokenType.Comment)
            {
                var trivia = tree.GetRoot().FindTrivia(expected.Position.Start);
                return (tree.GetLocation(trivia.FullSpan), TriviaClassifier.ClassifyTrivia(trivia));
            }
            else
            {
                var token = tree.GetRoot().FindToken(expected.Position.Start);
                var f = () => tokenClassifier.ClassifyToken(token);
                var tokenInfo = f.Should().NotThrow($"semanticModel should not be queried for {because}").Which;
                return (token.GetLocation(), tokenInfo);
            }
        }
```

In this way the methods are shorter, the level of imbrication is reduced and get easier to read/understand.